lib/trie/trie.go

```diff
@@ -333,11 +333,6 @@ func (t *Trie) Put(keyLE, value []byte) {
 
 func (t *Trie) insertKeyLE(keyLE, value []byte, deletedMerkleValues map[string]struct{}) {
 	nibblesKey := codec.KeyLEToNibbles(keyLE)
-	if len(value) == 0 {
-		// Force value to be inserted to nil since we don't
-		// differentiate between nil and empty values.
-		value = nil
-	}
```

Comment:
Sorry to block this; but all our code checks for empty values with `!= nil` and although that might the issue here, I don't want to risk other ramifications.

Especially:

- the trie node encoding would result in encoding empty node values as `{0}` for branch-without-value nodes, which is wrong according to spec and [substate code](https://github.com/paritytech/substrate/blob/a7ba55d3c54b9957c242f659e02f5b5a0f47b3ff/primitives/trie/src/node_codec.rs#L123). Our decoding would still work since we interpret `EOF` and `{0}` the same, but it's not necessarily the same with substrate. And it's an unneeded `{0}` as well.
- pretty much all trie methods rely on `SubValue != nil` checks

After re-reading Substrate code and the spec, #2927 is still correct and our code before was basically wrong. It was perhaps making it (still wrongly) thanks to the cached encoding of nodes, which got removed in #2919 

As much as I hate to throw the ball elsewhere, I think the problem you are facing is due to something else (and didn't show up because of our wrong node encoding caching before). I'll ask on the elements chat about node encoding/decoding just to double check though.